libraries-3/src/test/java/com/baeldung/jnats/NatsClientLiveTest.java

```diff
@@ -1,115 +1,202 @@
 
-        client.publishMessage("foo.bar", "bar.foo", "hello there");
+            Message message = qSub2.nextMessage(TIMEOUT_MILLIS);
+            if (message != null) {
+                messages.add(message);
+            }
 
-        Message message = fooSubscription.nextMessage(200);
-        assertNotNull("No message!", message);
-        assertEquals("hello there", new String(message.getData()));
+            message = qSub1.nextMessage(TIMEOUT_MILLIS);
+            if (message != null) {
+                messages.add(message);
+            }
+
+            assertEquals(1, messages.size());
+
+            client.unsubscribe(qSub1);
+            client.unsubscribe(qSub2);
+        }
     }
 
     @Test
-    public void whenWildCardSubscription_andNotMatchTopic_NoMessageReceived() throws Exception {
-
-        NatsClient client = connectClient();
+    public void whenSubscribingAsynchronouslyWithoutQueueGroups_thenEachMessageHandlerShouldReceiveEachMessage() throws Exception {
+        try (NatsClient client = connectClient()) {
+            List<Message> messages1 = new ArrayList<>();
+            List<Message> messages2 = new ArrayList<>();
 
-        SyncSubscription fooSubscription = client.subscribeSync("foo.*");
+            Subscription sub1 = client.subscribeAsync("mySubject", messages1::add);
+            Subscription sub2 = client.subscribeAsync("mySubject", messages2::add);
 
-        client.publishMessage("foo.bar.plop", "bar.foo", "hello there");
+            client.publishMessage("mySubject", "data1");
+            client.publishMessage("mySubject", "data2");
 
-        Message message = fooSubscription.nextMessage(200);
-        assertNull("Got message!", message);
+            Thread.sleep(WAIT_MILLIS); // give the messages time to be published and received
 
+            assertEquals(2, messages1.size());
+            assertEquals(2, messages2.size());
 
-        SyncSubscription barSubscription = client.subscribeSync("foo.>");
+            client.unsubscribe(sub1);
+            client.unsubscribe(sub2);
+        }
+    }
 
-        client.publishMessage("foo.bar.plop", "bar.foo", "hello there");
+    @Test
+    public void whenSubscribingAsynchronouslyWithQueueGroups_thenOnlyOneMessageHandlerInTheGroupShouldReceiveEachMessage() throws Exception {
+        try (NatsClient client = connectClient()) {
+            List<Message> messages = new ArrayList<>();
+            Subscription qSub1 = client.subscribeAsyncInQueueGroup("mySubject", "myQueue", messages::add);
+            Subscription qSub2 = client.subscribeAsyncInQueueGroup("mySubject", "myQueue", messages::add);
 
-        message = barSubscription.nextMessage(200);
-        assertNotNull("No message!", message);
-        assertEquals("hello there", new String(message.getData()));
+            client.publishMessage("mySubject", "data");
 
+            Thread.sleep(WAIT_MILLIS); // give the messages time to be published and received
```

Comment:
Let's avoid Thread.sleep as much as possible so the test is not flaky

```suggestion
 @Test
    public void whenSubscribingAsynchronouslyWithQueueGroups_thenOnlyOneMessageHandlerInTheGroupShouldReceiveEachMessage() throws Exception {
        try (NatsClient client = connectClient()) {
            CountDownLatch awaitMessages = new CountDownLatch(1);
            List<Message> messages = new ArrayList<>();
            MessageHandler add = message -> {
                awaitMessages.countDown();
                messages.add(message);
            };
            Subscription qSub1 = client.subscribeAsyncInQueueGroup("mySubject", "myQueue", add);
            Subscription qSub2 = client.subscribeAsyncInQueueGroup("mySubject", "myQueue", add);

            client.publishMessage("mySubject", "data");

            awaitMessages.await();
            Thread.sleep(WAIT_MILLIS); // wait a bit longer to prove only one message is received
            assertEquals(1, messages.size());

            client.unsubscribe(qSub1);
            client.unsubscribe(qSub2);
        }
    }
```

Reply:
So I understand the point of not using the sleep whenever possible, but it actually reduces the point of the test, because you are asserting a fact that you have forced by an external behavior in the test.

I think a better answer would be to clearly document why the sleep is being used.